File: src/csnav/viz/window_selector.py
"""A Leaflet control for isolating trajectory windows on a folium map.

A trajectory's manifest windows overlap by design - adjacent windows meet at a
shared boundary and each corridor is round-capped - so drawing them all at once
in one colour reads as a chain of blobs rather than a sequence of windows. The
fix is to put each window on its own layer and give the reader a way to pick
between them; folium's own ``LayerControl`` is flat, so listing 36 window
layers there would just move the problem.

:class:`WindowSelector` renders a collapsible tree instead: a row per
trajectory, expanding to that trajectory's windows, each with a checkbox and a
"solo" button that isolates it. Where a window has more than one kind of
geometry (its footprint, its candidate roads, its intersections, its imagery
tiles) those are **categories**, toggled across all windows at once, so a layer
is drawn when its window is selected *and* its category is enabled.

Layers handed to this control should be created with ``control=False`` so they
stay out of folium's own layer control - this one manages them.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Mapping, Sequence

from branca.element import MacroElement, Template

#: The control's behaviour, kept in a real ``.js`` file so it can be linted and
#: unit-tested under node rather than living inside a template string.
_SCRIPT_PATH = Path(__file__).parent / "static" / "window_selector.js"

_CSS = """
.csnav-ws {
  background: rgba(255, 255, 255, 0.94);
  padding: 6px 8px;
  border-radius: 4px;
  font: 12px/1.4 system-ui, -apple-system, "Segoe UI", Roboto, sans-serif;
  color: #222222;
  max-height: 60vh;
  overflow-y: auto;
  min-width: 190px;
  box-shadow: 0 1px 5px rgba(0, 0, 0, 0.4);
}
.csnav-ws-head { display: flex; align-items: center; gap: 4px; margin-bottom: 4px; }
.csnav-ws-title { font-weight: 600; flex: 1; }
.csnav-ws-cats {
  display: flex; flex-wrap: wrap; gap: 6px;
  padding: 3px 0 5px; margin-bottom: 4px; border-bottom: 1px solid #DDDDDD;
}
.csnav-ws-cat { display: flex; align-items: center; gap: 3px; cursor: pointer; white-space: nowrap; }
.csnav-ws-group { border-top: 1px solid #EEEEEE; }
.csnav-ws-grouphead { display: flex; align-items: center; gap: 4px; padding: 3px 0; cursor: pointer; }
.csnav-ws-caret { width: 9px; color: #666666; }
.csnav-ws-swatch { width: 9px; height: 9px; border-radius: 2px; flex: none; }
.csnav-ws-name { flex: 1; font-weight: 500; }
.csnav-ws-count { color: #666666; font-variant-numeric: tabular-nums; }
.csnav-ws-windows { display: none; padding: 0 0 4px 14px; }
.csnav-ws-window { display: flex; align-items: center; gap: 4px; padding: 1px 0; }
.csnav-ws-winlabel { display: flex; align-items: center; gap: 4px; flex: 1; cursor: pointer; }
.csnav-ws-wintext { font-variant-numeric: tabular-nums; }
.csnav-ws-btn {
  font: inherit; font-size: 10px; line-height: 1;
  padding: 2px 5px; cursor: pointer;
  border: 1px solid #CCCCCC; border-radius: 3px; background: #F7F7F7; color: #333333;
}
.csnav-ws-btn:hover { background: #E8E8E8; }
.csnav-ws input[type="checkbox"] { margin: 0; }
"""


@dataclass(frozen=True)
class WindowLayers:
    """One window's layers, keyed by category.

    ``window_id`` is the manifest window key (``"<trajectory_id>:0000"``);
    ``label`` is what the control shows; ``detail`` becomes the row's tooltip.
    ``layers`` maps a category key to the folium layer holding that category's
    geometry for this window - typically a ``FeatureGroup`` built with
    ``control=False``.
    """

    window_id: str
    label: str
    layers: Mapping[str, Any]
    detail: str = ""


@dataclass(frozen=True)
class WindowGroup:
    """A trajectory's windows, as one collapsible row of the control."""

    id: str
    label: str
    color: str
    windows: tuple[WindowLayers, ...] = field(default_factory=tuple)


@dataclass(frozen=True)
class WindowCategory:
    """A kind of geometry that exists once per window, toggled across all of them.

    ``enabled`` is the initial state - set it ``False`` for a category that is
    useful but noisy by default, like every window's imagery tiles.
    """

    key: str
    label: str
    enabled: bool = True


class WindowSelector(MacroElement):
    """The collapsible per-window layer control.

    ``groups`` are the trajectories to list; ``categories`` the kinds of
    geometry each window carries (omit for a single unnamed category, in which
    case no category row is drawn). ``title`` heads the panel and ``position``
    is any Leaflet control position.

    Add it to the map *after* the layers it manages: it references their
    JavaScript variables directly, so they have to be declared first.
    """

    _template = Template(
        """
        {% macro header(this, kwargs) %}
            <style>{{ this.css }}</style>
        {% endmacro %}

        {% macro script(this, kwargs) %}
            {{ this.script }}
            csnavWindowSelector({{ this._parent.get_name() }}, {{ this.spec }}, {{ this.options }});
        {% endmacro %}
        """
    )

    def __init__(
        self,
        groups: Sequence[WindowGroup],
        categories: Sequence[WindowCategory] = (),
        title: str = "Windows",
        position: str = "topright",
    ) -> None:
        super().__init__()
        self._name = "WindowSelector"
        self.groups = list(groups)
        self.categories = list(categories)
        self.title = title
        self.position = position
        self.css = _CSS
        self.script = _SCRIPT_PATH.read_text(encoding="utf-8")

    @property
    def options(self) -> str:
        return "{position: %s}" % _js_string(self.position)

    @property
    def spec(self) -> str:
        """The control's spec as a JavaScript object literal.

        Not JSON: layer references have to come out as the bare JavaScript
        variable names folium declared for them, so they resolve to the actual
        Leaflet layers rather than to strings. Every piece of text goes through
        :func:`_js_string`, so a label cannot break out of the literal - or out
        of the ``<script>`` element the literal is written into.
        """
        categories = ", ".join(
            "{key: %s, label: %s, enabled: %s}"
            % (_js_string(category.key), _js_string(category.label), _js_bool(category.enabled))
            for category in self.categories
        )
        groups = ", ".join(self._group_literal(group) for group in self.groups)
        return (
            "{title: %s, categories: [%s], groups: [%s]}"
            % (_js_string(self.title), categories, groups)
        )

    def _group_literal(self, group: WindowGroup) -> str:
        windows = ", ".join(self._window_literal(window) for window in group.windows)
        return "{id: %s, label: %s, color: %s, windows: [%s]}" % (
            _js_string(group.id),
            _js_string(group.label),
            _js_string(group.color),
            windows,
        )

    def _window_literal(self, window: WindowLayers) -> str:
        layers = ", ".join(
            "%s: %s" % (_js_string(key), layer.get_name()) for key, layer in window.layers.items()
        )
        return "{id: %s, label: %s, detail: %s, layers: {%s}}" % (
            _js_string(window.window_id),
            _js_string(window.label),
            _js_string(window.detail),
            layers,
        )


def _js_bool(value: bool) -> str:
    return "true" if value else "false"


def _js_string(value: str) -> str:
    r"""A JavaScript string literal that is safe inside an inline ``<script>`` element.

    ``json.dumps`` alone is not enough: it escapes quotes and backslashes but
    leaves ``<`` untouched, so a label containing ``</script>`` would close the
    element early and everything after it would be parsed as markup. Escaping
    the three markup-significant characters as ``\uXXXX`` keeps the literal
    inert while decoding back to the original text at runtime.
    """
    escaped = json.dumps(value)
    for character, replacement in (("<", r"\u003c"), (">", r"\u003e"), ("&", r"\u0026")):
        escaped = escaped.replace(character, replacement)
    return escaped

File: src/csnav/viz/static/window_selector.js
/*
 * Window selector: a Leaflet control for isolating trajectory windows.
 *
 * A trajectory's manifest windows overlap by design - adjacent windows share a
 * boundary and each is round-capped - so drawing them all at once in one colour
 * produces a chain of blobs rather than a readable sequence. This control puts
 * each window on its own layer and gives you a tree to pick from: expand a
 * trajectory, tick the windows you want, or "solo" one to see it alone.
 *
 * Layers are addressed as a (window x category) matrix. A layer is on the map
 * when its window is selected AND its category is enabled, so "show every
 * window's roads but no footprints" and "show only window 3, everything about
 * it" are both reachable.
 *
 * The state half of this file is deliberately free of DOM and Leaflet calls so
 * it can be unit-tested under node; see tests/viz/test_window_selector.js.
 */

/**
 * Selection state for a window-selector spec.
 *
 * spec: {categories: [{key, label, enabled}], groups: [{id, label, color,
 *        windows: [{id, label, detail, layers: {categoryKey: <leaflet layer>}}]}]}
 */
function csnavWindowSelectorState(spec) {
  var windowOn = {};
  var categoryOn = {};
  var groups = spec.groups || [];
  var categories = spec.categories || [];

  groups.forEach(function (group) {
    (group.windows || []).forEach(function (win) {
      windowOn[win.id] = true;
    });
  });
  categories.forEach(function (category) {
    categoryOn[category.key] = category.enabled !== false;
  });

  function windowsOf(groupId) {
    var found = groups.filter(function (group) {
      return group.id === groupId;
    })[0];
    return found ? found.windows || [] : [];
  }

  function allWindows() {
    return groups.reduce(function (acc, group) {
      return acc.concat(group.windows || []);
    }, []);
  }

  return {
    /** Whether one window's layers should be drawn at all. */
    isWindowOn: function (windowId) {
      return windowOn[windowId] === true;
    },
    /** Whether a category is enabled across every window. */
    isCategoryOn: function (categoryKey) {
      return categoryOn[categoryKey] === true;
    },
    /** Whether the layer for (window, category) belongs on the map right now. */
    isVisible: function (windowId, categoryKey) {
      return windowOn[windowId] === true && categoryOn[categoryKey] === true;
    },
    setWindow: function (windowId, on) {
      windowOn[windowId] = on === true;
    },
    setCategory: function (categoryKey, on) {
      categoryOn[categoryKey] = on === true;
    },
    /** Turn every window in one group on or off. */
    setGroup: function (groupId, on) {
      windowsOf(groupId).forEach(function (win) {
        windowOn[win.id] = on === true;
      });
    },
    /** Show one window and nothing else - the isolate action. */
    solo: function (windowId) {
      allWindows().forEach(function (win) {
        windowOn[win.id] = win.id === windowId;
      });
    },
    setAll: function (on) {
      allWindows().forEach(function (win) {
        windowOn[win.id] = on === true;
      });
    },
    /** How many of a group's windows are currently on. */
    countOn: function (groupId) {
      return windowsOf(groupId).filter(function (win) {
        return windowOn[win.id] === true;
      }).length;
    },
    groups: groups,
    categories: categories
  };
}

/* Rendering below this line: DOM and Leaflet only. */

function csnavWindowSelector(map, spec, options) {
  var settings = options || {};
  var state = csnavWindowSelectorState(spec);
  var expanded = {};
  var rendered = null;

  function element(tag, className, text) {
    var node = document.createElement(tag);
    if (className) {
      node.className = className;
    }
    if (text !== undefined) {
      node.textContent = text;
    }
    return node;
  }

  function applyLayers() {
    state.groups.forEach(function (group) {
      (group.windows || []).forEach(function (win) {
        Object.keys(win.layers || {}).forEach(function (key) {
          var layer = win.layers[key];
          if (!layer) {
            return;
          }
          if (state.isVisible(win.id, key)) {
            if (!map.hasLayer(layer)) {
              map.addLayer(layer);
            }
          } else if (map.hasLayer(layer)) {
            map.removeLayer(layer);
          }
        });
      });
    });
  }

  function refresh() {
    applyLayers();
    if (rendered) {
      rendered();
    }
  }

  function buildCategories(container) {
    if (state.categories.length < 2) {
      return function () {};
    }
    var row = element('div', 'csnav-ws-cats');
    var inputs = [];
    state.categories.forEach(function (category) {
      var label = element('label', 'csnav-ws-cat');
      var box = document.createElement('input');
      box.type = 'checkbox';
      box.checked = state.isCategoryOn(category.key);
      box.addEventListener('change', function () {
        state.setCategory(category.key, box.checked);
        refresh();
      });
      label.appendChild(box);
      label.appendChild(element('span', null, category.label));
      row.appendChild(label);
      inputs.push({ box: box, key: category.key });
    });
    container.appendChild(row);
    return function () {
      inputs.forEach(function (entry) {
        entry.box.checked = state.isCategoryOn(entry.key);
      });
    };
  }

  function buildGroup(group, container) {
    var block = element('div', 'csnav-ws-group');
    var head = element('div', 'csnav-ws-grouphead');
    var caret = element('span', 'csnav-ws-caret', '▸');
    var swatch = element('span', 'csnav-ws-swatch');
    swatch.style.background = group.color || '#666666';
    var name = element('span', 'csnav-ws-name', group.label || group.id);
    var count = element('span', 'csnav-ws-count');

    head.appendChild(caret);
    head.appendChild(swatch);
    head.appendChild(name);
    head.appendChild(count);

    ['all', 'none'].forEach(function (action) {
      var button = element('button', 'csnav-ws-btn', action);
      button.type = 'button';
      button.title = action === 'all' ? 'show every window' : 'hide every window';
      button.addEventListener('click', function (event) {
        event.stopPropagation();
        state.setGroup(group.id, action === 'all');
        refresh();
      });
      head.appendChild(button);
    });

    var list = element('div', 'csnav-ws-windows');
    var boxes = [];
    (group.windows || []).forEach(function (win) {
      var row = element('div', 'csnav-ws-window');
      var label = element('label', 'csnav-ws-winlabel');
      var box = document.createElement('input');
      box.type = 'checkbox';
      box.checked = state.isWindowOn(win.id);
      box.addEventListener('change', function () {
        state.setWindow(win.id, box.checked);
        refresh();
      });
      label.appendChild(box);
      label.appendChild(element('span', 'csnav-ws-wintext', win.label));
      if (win.detail) {
        label.title = win.detail;
      }
      row.appendChild(label);

      var solo = element('button', 'csnav-ws-btn', 'solo');
      solo.type = 'button';
      solo.title = 'show only this window';
      solo.addEventListener('click', function (event) {
        event.stopPropagation();
        state.solo(win.id);
        refresh();
      });
      row.appendChild(solo);
      list.appendChild(row);
      boxes.push({ box: box, id: win.id });
    });

    head.addEventListener('click', function () {
      expanded[group.id] = !expanded[group.id];
      refresh();
    });

    block.appendChild(head);
    block.appendChild(list);
    container.appendChild(block);

    return function () {
      var total = (group.windows || []).length;
      count.textContent = state.countOn(group.id) + '/' + total;
      caret.textContent = expanded[group.id] ? '▾' : '▸';
      list.style.display = expanded[group.id] ? 'block' : 'none';
      boxes.forEach(function (entry) {
        entry.box.checked = state.isWindowOn(entry.id);
      });
    };
  }

  var Control = L.Control.extend({
    onAdd: function () {
      var root = element('div', 'csnav-ws leaflet-bar');
      var head = element('div', 'csnav-ws-head');
      head.appendChild(element('span', 'csnav-ws-title', spec.title || 'Windows'));

      ['all', 'none'].forEach(function (action) {
        var button = element('button', 'csnav-ws-btn', action);
        button.type = 'button';
        button.title = action === 'all' ? 'show every window' : 'hide every window';
        button.addEventListener('click', function () {
          state.setAll(action === 'all');
          refresh();
        });
        head.appendChild(button);
      });
      root.appendChild(head);

      var refreshCategories = buildCategories(root);
      var body = element('div', 'csnav-ws-body');
      var refreshers = state.groups.map(function (group) {
        return buildGroup(group, body);
      });
      root.appendChild(body);

      rendered = function () {
        refreshCategories();
        refreshers.forEach(function (fn) {
          fn();
        });
      };

      L.DomEvent.disableClickPropagation(root);
      L.DomEvent.disableScrollPropagation(root);
      refresh();
      return root;
    }
  });

  var control = new Control({ position: settings.position || 'topright' });
  control.addTo(map);
  if (spec.groups && spec.groups.length === 1) {
    expanded[spec.groups[0].id] = true;
    refresh();
  }
  return control;
}

/* Exported for the node-based unit tests; `module` is undefined in a browser. */
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { csnavWindowSelectorState: csnavWindowSelectorState };
}
